### PR TITLE
Admin controller user delete

### PR DESCRIPTION
**Pull request by prateeksingh**
Added Delete User api., please review.

### CODE REVIEW

File: quora-api/src/main/java/com/upgrad/quora/api/controller/AdminController.java [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+import com.upgrad.quora.service.exception.AuthorizationFailedException;
+import com.upgrad.quora.service.exception.UserNotFoundException;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.http.HttpStatus;
+import org.springframework.http.ResponseEntity;
+import org.springframework.web.bind.annotation.*;
+
+
+@RestController
+@RequestMapping("/admin")
```
Comment (parimalvfx) [inline quora-api/src/main/java/com/upgrad/quora/api/controller/AdminController.java]: Should it be?
```suggestion
@RequestMapping("/")`
```

File: quora-service/src/main/java/com/upgrad/quora/service/entity/UserEntity.java [hunk 2]
```diff
@@ -12,7 +14,10 @@
 @NamedQueries(
         {
                 @NamedQuery(name = "userByUserName", query = "select u from UserEntity u where u.userName = :userName"),
-                @NamedQuery(name = "userByEmail", query = "select u from UserEntity u where u.email = :email")
+                @NamedQuery(name = "userByEmail", query = "select u from UserEntity u where u.email = :email"),
+                @NamedQuery(name = "userById", query = "select u from UserEntity u where u.uuid = :uid"),
```
Comment (parimalvfx) [inline quora-service/src/main/java/com/upgrad/quora/service/entity/UserEntity.java]:
```suggestion
                @NamedQuery(name = "userById", query = "select u from UserEntity u where u.id = :id"),
```

File: quora-service/src/main/java/com/upgrad/quora/service/dao/UserDao.java [hunk 3]
```diff
@@ -55,6 +55,33 @@ public UserEntity getUserByEmail(final String email) {
             return null;
         }
     }
+    /**
+     * This method helps find existing user by email id
```
Comment (parimalvfx) [inline quora-service/src/main/java/com/upgrad/quora/service/dao/UserDao.java]:
```suggestion
     * This method helps find existing user by uuid
```

File: quora-service/src/main/java/com/upgrad/quora/service/dao/UserDao.java [hunk 4]
```diff
@@ -55,6 +55,33 @@ public UserEntity getUserByEmail(final String email) {
...
+     *
+     * @param uid the user id which will be searched in database for existing user
```
Comment (parimalvfx) [inline quora-service/src/main/java/com/upgrad/quora/service/dao/UserDao.java]:
```suggestion
     * @param uuid the user id which will be searched in database for existing user
```

File: quora-service/src/main/java/com/upgrad/quora/service/dao/UserDao.java [hunk 5]
```diff
@@ -55,6 +55,33 @@ public UserEntity getUserByEmail(final String email) {
...
+     *
+     * @return UserEntity object if user with requested email id exists in database
```
Comment (parimalvfx) [inline quora-service/src/main/java/com/upgrad/quora/service/dao/UserDao.java]:
```suggestion
     * @return UserEntity object if user with requested uuid exists in database
```

File: quora-service/src/main/java/com/upgrad/quora/service/dao/UserDao.java [hunk 6]
```diff
@@ -55,6 +55,33 @@ public UserEntity getUserByEmail(final String email) {
...
+     */
+    public UserEntity getUserById(String uid){
```
Comment (parimalvfx) [inline quora-service/src/main/java/com/upgrad/quora/service/dao/UserDao.java]:
```suggestion
    public UserEntity getUserByUuid(final String uuid){
```

File: quora-service/src/main/java/com/upgrad/quora/service/dao/UserDao.java [hunk 7]
```diff
@@ -55,6 +55,33 @@ public UserEntity getUserByEmail(final String email) {
...
+
+        try{
+            return entityManager.createNamedQuery("userById",UserEntity.class).setParameter("uuid",uid).getSingleResult();
```
Comment (parimalvfx) [inline quora-service/src/main/java/com/upgrad/quora/service/dao/UserDao.java]:
```suggestion
            return entityManager.createNamedQuery("userByUuid", UserEntity.class).setParameter("uuid", uuid).getSingleResult();
```

File: quora-service/src/main/java/com/upgrad/quora/service/dao/UserDao.java [hunk 8]
```diff
@@ -55,6 +55,33 @@ public UserEntity getUserByEmail(final String email) {
...
+        } catch (NoResultException nre){
+            return null;
+        }
+
+
+    }
+
+
+    public UserEntity deleteUser(String uuid) {
```
Comment (parimalvfx) [inline quora-service/src/main/java/com/upgrad/quora/service/dao/UserDao.java]:
```suggestion
    public UserEntity deleteUser(final String uuid) {
```

File: quora-service/src/main/java/com/upgrad/quora/service/entity/UserEntity.java [hunk 9]
```diff
@@ -12,7 +14,10 @@
...
+                @NamedQuery(name = "deleteUserById", query="delete from UserEntity u where u.uuid = :uid"),
```
Comment (parimalvfx) [inline quora-service/src/main/java/com/upgrad/quora/service/entity/UserEntity.java]:
```suggestion
                @NamedQuery(name = "deleteUserByUuid", query="delete from UserEntity u where u.uuid = :uuid"),
```

File: quora-service/src/main/java/com/upgrad/quora/service/dao/UserDao.java [hunk 10]
```diff
@@ -55,6 +55,33 @@ public UserEntity getUserByEmail(final String email) {
...
+
+        try{
+            return entityManager.createNamedQuery("deleteUserById",UserEntity.class).setParameter("uuid",uuid).getSingleResult();
```
Comment (parimalvfx) [inline quora-service/src/main/java/com/upgrad/quora/service/dao/UserDao.java]:
```suggestion
            return entityManager.createNamedQuery("deleteUserByUuid",UserEntity.class).setParameter("uuid", uuid).getSingleResult();
```

File: quora-service/src/main/java/com/upgrad/quora/service/business/AdminBusinessService.java [hunk 11]
```diff
@@ -72,4 +74,30 @@ public UserEntity logoutUser(final String authorizationToken) throws SignOutRest
 
         return userAuthEntity.getUser();
     }
+
+    public void deleteUser(final String uuid, final String authorizationToken) throws UserNotFoundException, AuthorizationFailedException {
```
Comment (parimalvfx) [inline quora-service/src/main/java/com/upgrad/quora/service/business/AdminBusinessService.java]: Exception handler not added for `UserNotFoundException`. Please add it in `com.upgrad.quora.api.exception.RestExceptionHandler` class.

File: quora-service/src/main/java/com/upgrad/quora/service/business/AdminBusinessService.java [hunk 12]
```diff
@@ -72,4 +74,30 @@ public UserEntity logoutUser(final String authorizationToken) throws SignOutRest
...
+
+        //UserEntity existingUser = userDao.getUserByUserName(userEntity.getUserName());
```
Comment (parimalvfx) [inline quora-service/src/main/java/com/upgrad/quora/service/business/AdminBusinessService.java]: Let's remove commented code.

File: quora-service/src/main/java/com/upgrad/quora/service/business/AdminBusinessService.java [hunk 13]
```diff
@@ -72,4 +74,30 @@ public UserEntity logoutUser(final String authorizationToken) throws SignOutRest
...
+
+        if (userDao.getUserById(uuid)==null) {
```
Comment (parimalvfx) [inline quora-service/src/main/java/com/upgrad/quora/service/business/AdminBusinessService.java]: 1. We should have first validation for user authorization then for request details.

2. Code change:
```suggestion
           if (userDao.getUserByUuid(uuid) == null) {
```

File: quora-api/src/main/java/com/upgrad/quora/api/controller/AdminController.java [hunk 14]
```diff
@@ -0,0 +1,31 @@
...
+public class AdminController {
+
+    @Autowired
+    private AdminBusinessService adminBusinessService;
+
+    @RequestMapping(method = RequestMethod.DELETE, value="/admin/user/{userId}")
+    public ResponseEntity<UserDeleteResponse> deleteUser(@PathVariable String userId, @RequestHeader("authorization") final String authorization) throws AuthorizationFailedException, UserNotFoundException {
```
Comment (parimalvfx) [inline quora-api/src/main/java/com/upgrad/quora/api/controller/AdminController.java]:
```suggestion
    public ResponseEntity<UserDeleteResponse> deleteUser(@PathVariable("userId") final String userId, @RequestHeader("authorization") final String authorization) throws AuthorizationFailedException, UserNotFoundException {
```

**Comment by prateeksingh** [comment]
@parimalvfx  Done all the changes as you suggested

**Comment by parimalvfx** [comment]
Issue #5

**Pull request merged**